tests/std/tests/GH_004597_self_swap/test.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <utility>
+
+struct swap_counter {
+    unsigned int* pcnt_ = nullptr;
+
+    friend _CONSTEXPR20 void swap(swap_counter& lhs, swap_counter& rhs) noexcept {
+        std::swap(lhs.pcnt_, rhs.pcnt_);
+        if (lhs.pcnt_ != nullptr) {
+            ++(*lhs.pcnt_);
+        }
+        if (rhs.pcnt_ != nullptr) {
+            ++(*rhs.pcnt_);
+        }
+    }
+
+    _CONSTEXPR20 bool operator==(unsigned int x) const {
+        return *pcnt_ == x;
+    }
+};
+
+
+_CONSTEXPR20 bool test_gh_4595() {
+    auto res1 = [] {
+        unsigned int cnt{};
+        std::pair<swap_counter, int> pr{swap_counter{&cnt}, 0};
+        pr.swap(pr);
+        return (cnt == 2u) && (pr.first == 2u) && (pr.second == 0);
+    }();
+
+
+    auto res2 = [] {
+        unsigned int cnt{};
+        std::pair<swap_counter, int> p1{swap_counter{&cnt}, 0};
+        std::pair<swap_counter, int> p2{swap_counter{&cnt}, 1};
+        p1.swap(p2);
+        return (cnt == 2u) && (p1.first == 2u) && (p1.second == 1) && (p2.first == 2u) && (p2.second == 0);
+    }();
+
+
+    auto res3 = [] {
+        unsigned int c1{};
+        unsigned int c2{2};
+        std::pair<swap_counter, int> p1{swap_counter{&c1}, 1};
+        std::pair<swap_counter, int> p2{swap_counter{&c2}, 3};
+        p1.swap(p2);
+        return (c1 == 1u) && (c2 == 3u) && (p1.first == 3u) && (p1.second == 3) && (p2.first == 1u) && (p2.second == 1);
+    }();
+
+    auto res4 = true;
+
+#if _HAS_CXX23
+    res4 = [] {
+        unsigned int c1{};
+        unsigned int c2{2};
+        int i1  = 1;
+        int i2  = 3;

```

Comment:
This is a test style nitpick, but I consider it to be fairly important as far as nitpicks go: when writing tests, I strongly recommend "unique numbers for unique purposes". Here, you have two distinct kinds of integers - swap counters (the `first` elements) and plain `int` payloads (the `second` elements). There's no reason to reuse the same numbers, and it makes it a bit harder to follow along where the `c2` counter is increased from `2u` to `3u`, when one of the `int`s is also `3`.